glean-core/rlb/src/test.rs

```diff
@@ -171,15 +171,143 @@ fn test_experiments_recording_before_glean_inits() {
 }
 
 #[test]
-#[ignore] // TODO: To be done in bug 1673645.
-fn test_sending_of_foreground_background_pings() {
-    todo!()
+fn sending_of_foreground_background_pings() {
+    let _lock = lock_test();
+
+    let click: EventMetric<traits::NoExtraKeys> = private::EventMetric::new(CommonMetricData {
+        name: "click".into(),
+        category: "ui".into(),
+        send_in_pings: vec!["events".into()],
+        lifetime: Lifetime::Ping,
+        disabled: false,
+        ..Default::default()
+    });
+
+    // Define a fake uploader that reports back the submission headers
+    // using a crossbeam channel.
+    let (s, r) = crossbeam_channel::bounded::<String>(3);
+
+    #[derive(Debug)]
+    pub struct FakeUploader {
+        sender: crossbeam_channel::Sender<String>,
+    }
+    impl net::PingUploader for FakeUploader {
+        fn upload(
+            &self,
+            url: String,
+            _body: Vec<u8>,
+            _headers: Vec<(String, String)>,
+        ) -> net::UploadResult {
+            self.sender.send(url).unwrap();
+            net::UploadResult::HttpStatus(200)
+        }
+    }
+
+    // Create a custom configuration to use a fake uploader.
+    let dir = tempfile::tempdir().unwrap();
+    let tmpname = dir.path().display().to_string();
+
+    let cfg = Configuration {
+        data_path: tmpname,
+        application_id: GLOBAL_APPLICATION_ID.into(),
+        upload_enabled: true,
+        max_events: None,
+        delay_ping_lifetime_io: false,
+        channel: Some("testing".into()),
+        server_endpoint: Some("invalid-test-host".into()),
+        uploader: Some(Box::new(FakeUploader { sender: s })),
+    };
+
+    let _t = new_glean(Some(cfg), true);
+    crate::block_on_dispatcher();
+
+    // Simulate becoming active.
+    handle_client_active();
+
+    // We expect a baseline ping to be generated here (reason: 'active').
+    let url = r.recv().unwrap();
+    assert!(url.contains("baseline"));
+
+    // Recording an event so that an "events" ping will contain data.
+    click.record(None);
+
+    // Simulate becoming inactive
+    handle_client_inactive();
+
+    // Wait for the pings to arrive.
+    let mut expected_pings = vec!["baseline", "events"];
+    for _ in 0..2 {
+        let url = r.recv().unwrap();
+        // If the url contains the expected reason, remove it from the list.
+        expected_pings.retain(|&name| url.contains(name));
```

Comment:
This doesn't actually does what we want, does it?
We want to check that we receive the 2 expected pings and no other.
Right now we check that ... barely anything.
Easy to fix:

```suggestion
        expected_pings.retain(|&name| !url.contains(name));
```

Longer details: 

<details>

`retain` removes elements when the closure returns `false`.
If the first ping that gets in is `baseline`, then:
1. it checks if the url contains `baseline`. This is true. The element is kept.
2. it checks if the url contains `events`. This is false. `events` is removed.
We're left with `expected_pings = ["baseline"]`
If we would now get a `deletion-request` ping, then:
1. it checks if the url contains `baseline`. This is false. `baseline` is removed.
We're left with `expected_pings = []`

By inverting the retain check we keep everything that's not ping we currently look at.
</details>

Reply:
good catch!